compiler/rustc_mir_transform/src/dataflow_const_prop.rs

```diff
@@ -540,12 +540,114 @@ impl<'tcx, 'locals> CollectAndPatch<'tcx, 'locals> {
         }
     }
 
-    fn make_operand(&self, scalar: ScalarInt, ty: Ty<'tcx>) -> Operand<'tcx> {
-        Operand::Constant(Box::new(Constant {
-            span: DUMMY_SP,
-            user_ty: None,
-            literal: ConstantKind::Val(ConstValue::Scalar(scalar.into()), ty),
-        }))
+    fn try_make_constant(
+        &self,
+        place: Place<'tcx>,
+        state: &State<FlatSet<Scalar>>,
+        map: &Map,
+    ) -> Option<ConstantKind<'tcx>> {
+        let ty = place.ty(self.local_decls, self.tcx).ty;
+        let place = map.find(place.as_ref())?;
+        if let FlatSet::Elem(Scalar::Int(value)) = state.get_idx(place, map) {
+            Some(ConstantKind::Val(ConstValue::Scalar(value.into()), ty))
+        } else {
+            let valtree = self.try_make_valtree(place, ty, state, map)?;
+            let constant = ty::Const::new_value(self.tcx, valtree, ty);
+            Some(ConstantKind::Ty(constant))
+        }
+    }
+
+    fn try_make_valtree(
```

Comment:
I need to duplicate this method into a `check_` version and a `make_` version. The `try_make_` version leaks memory if it fails to complete the ValTree, which is more than likely in general.

Reply:
Hmm... Maybe we should have some sort of `OwnedValtree` that can infallibly be converted to a `Valtree`. This way you can create the `OwnedValtree` and dropping it frees all memory?